frontend_csv/src/parser/parser.rs

```diff
@@ -0,0 +1,230 @@
+//! Core Parser implementation for the csv tokens
+use std::collections::BTreeMap;
+use std::vec::Vec;
+
+use crate::parser::ast::LNMsData;
+use crate::parser::ast::LNMsg;
+use crate::parser::ast::LNTlvData;
+use crate::parser::ast::LNTlvType;
+use crate::scanner::token::{CSVToken, CSVTokenType};
+
+pub struct Parser {
+    pub simple_table: BTreeMap<String, LNMsg>,
+    pos: usize,
+    lnmsg_buffer: Option<LNMsg>,
+    lntlv_buffer: Option<LNTlvType>,
+}
+
+impl Parser {
+    /// Build a new parser
+    pub fn new() -> Self {
+        return Parser {
+            pos: 0,
+            simple_table: BTreeMap::new(),
+            lnmsg_buffer: None,
+            lntlv_buffer: None,
+        };
+    }
+
+    /// Take the element in the current position of the stream
+    fn peek<'a>(&self, tokens: &'a Vec<CSVToken>) -> &'a CSVToken {
+        return &tokens[self.pos];
+    }
+
+    /// Take the element in the current position of the stream
+    /// and increase the position by one
+    fn pop(&mut self, tokens: &Vec<CSVToken>) -> CSVToken {
+        let res = tokens[self.pos].clone();
+        self.pos += 1;
+        return res;
+    }
```

Comment:
```suggestion
    fn pop<'a>(&mut self, tokens: &Vec<CSVToken>) -> &'a CSVToken {
        self.pos += 1;
        return &tokens[self.pos];
    }
```

Clone is not always the solution :)

Reply:
Hi I try to use 'a for tokens but it won't work as the token returned is changed

Reply:
what is that mean?

This diff works

```diff
diff --git a/frontend_csv/src/parser/parser.rs b/frontend_csv/src/parser/parser.rs
index 7a09c79..acd97cb 100644
--- a/frontend_csv/src/parser/parser.rs
+++ b/frontend_csv/src/parser/parser.rs
@@ -33,10 +33,9 @@ impl Parser {
 
     /// Take the element in the current position of the stream
     /// and increase the position by one
-    fn advance(&mut self, tokens: &Vec<CSVToken>) -> CSVToken {
-        let res = tokens[self.pos].clone();
+    fn advance<'a>(&mut self, tokens: &'a Vec<CSVToken>) -> &'a CSVToken {
         self.pos += 1;
-        return res;
+        return &tokens[self.pos];
     }
 
     /// Parse a message type line of the csv file, where the format looks like
@@ -59,10 +58,8 @@ impl Parser {
         }
     }
 
-
-
     // This method is to get the lenght of the integer type for bytes field
-    // for example: 
+    // for example:
     // msgdata,init,flen,u16,
     // msgdata,init,features,byte,flen
     // flen is the length of the bytes field which is the type u16 which is 2 bytes
@@ -111,11 +108,14 @@ impl Parser {
             CSVTokenType::U32 => self.add_data_lnmsg_buffer(LNMsData::Unsigned32(msg_data_name, 4)),
             CSVTokenType::U64 => self.add_data_lnmsg_buffer(LNMsData::Unsigned64(msg_data_name, 8)),
             CSVTokenType::ChainHash => {
-                let msg_val = self.advance(&tokens).val;
-                self.add_data_lnmsg_buffer(LNMsData::ChainHash(msg_data_name, msg_val));
+                let msg_val = &self.advance(&tokens);
+                self.add_data_lnmsg_buffer(LNMsData::ChainHash(
+                    msg_data_name,
+                    msg_val.val.to_string(),
+                ));
             }
             CSVTokenType::Byte => {
-                let byte_name = self.advance(&tokens).val.to_string();
+                let byte_name = &self.advance(&tokens).val.to_string();
                 print!("\n bytes name {:?}\n", byte_name);
                 let msg_val = self.get_byte_length(byte_name.clone());
                 self.add_data_lnmsg_buffer(LNMsData::BitfieldStream(
```

Reply:
@michiboo pushed the change! it works as you can see